sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/Exchange.scala

```diff
@@ -109,3 +109,67 @@ case class ReuseExchange(conf: SQLConf) extends Rule[SparkPlan] {
     }
   }
 }
+
+/**
+ * Find out duplicated coordinated exchanges in the spark plan, then use the same exchange for all
+ * the references.
+ */
+case class ReuseExchangeWithCoordinator(conf: SQLConf) extends Rule[SparkPlan] {
+
+  // Returns true if a SparkPlan has coordinated ShuffleExchangeExec children.
+  private def hasCoordinatedExchanges(plan: SparkPlan): Boolean = {
+    plan.children.nonEmpty && plan.children.forall {
+      case ShuffleExchangeExec(_, _, Some(_)) => true
+      case _ => false
+    }
+  }
+
+  // Returns true if two sequences of exchanges are producing the same results.
+  private def hasExchangesWithSameResults(
+      source: Seq[ShuffleExchangeExec],
+      target: Seq[ShuffleExchangeExec]): Boolean = {
+    source.length == target.length &&
+      source.zip(target).forall(x => x._1.withoutCoordinator.sameResult(x._2.withoutCoordinator))
```

Comment:
Actually I think this can be relaxed to:

```scala
val sourceWithoutCoordinator = source.map(_.withoutCoordinator)
val targetWithoutCoordinator = target.map(_.withoutCoordinator)
source.length == target.length && sourceWithoutCoordinator.forall(s => targetWithoutCoordinator.exists(_.sameResult(s)))
```

For simplicity and safety, I leave it as this.